### ISSUE TITLE
Local changes clobbered during concurrent saves

### ISSUE THREAD

**Issue by mmun** [emberjs/data#4195]
There seems to be some unusual behaviour when making multiple `.save()` calls concurrently. Specifically, if the PUT response contains data (e.g. the record that was just saved) it will clobber any existing local changes.

Here is an ember-twiddle showing the issue with steps to reproduce:

~~https://ember-twiddle.com/6bc630a36d36aedde158~~ - ember-twiddle changed something and the demo no longer loads correctly.

Fixed by Clemens: https://ember-twiddle.com/007ad961737ab7caa258f5c18d6c9253?openFiles=application.controller.js%2C

**Comment by fivetanley** [comment]
So is the desired behavior to cancel any previous pending requests when a new request is made?

**Comment by fivetanley** [comment]
Or just to not clobber any local changes?

**Comment by mmun** [comment]
I'm not too sure but my gut reaction was the latter option.

**Comment by pangratz** [comment]
Related #4144

**Comment by workmanw** [comment]
@pangratz I think at least [my comment](https://github.com/emberjs/data/issues/4144#issuecomment-182119676) in #4144 is relevant. We had this problem in our app and ultimately we just created a "save throttling mechanism".  Here is a gist: https://gist.github.com/workmanw/aa02325ea92fdd4744cb . 

From my experience with this I think both problems, an error and data clobbering, can happen depending on the sequence of events (`set`, `save` and `didCommit`). Very racey.

**Comment by pangratz** [comment]
@workmanw you're right, both issues are in the same problem space.

**Comment by wecc** [comment]
@mmun @fivetanley @pangratz are you OK with me closing this as a dupe of #4144 to make it easier to track? I interpret both issues as the same problem.

**Comment by mmun** [comment]
@wecc I'm not sure they're the same thing. Unlike #4144, there are no errors thrown when exhibiting this issue. It is just an unexpected behaviour (local changes are clobbered by the remote response).

**Comment by pangratz** [comment]
[Working version of initial ember-twiddle which seems to not be working anymore](https://ember-twiddle.com/007ad961737ab7caa258f5c18d6c9253?fileTreeShown=false&openFiles=twiddle.json%2C).

**Comment by mmun** [comment]
Thank you @pangratz !

**Comment by robbytx** [comment]
FWIW this is still an issue on Ember Data 4.7.  Concurrent `.save()` invocations for the same record may be mishandled because the data returned by the first `.save()` is taken as the canonical set of attributes and the changes in the second `.save()` are forgotten, unless and until the second save resolves.

Example:
```js
// 1. First update
record.counter = 1;
const save1 = record.save();

// 2. Second update (issued in separate runloop, but before save1 resolves):
record.counter = 2;
const save2 = record.save();

// 3. After save1 resolves but while the concurrent save2 request is still in-flight:
(await save1).counter === 1;
// At this point, it looks like counter is 1 and the update to 2 was effectively lost (except for the pending save2 promise).

// 4. Eventually save2 completes and its response takes effect.
(await save2).counter === 2;
```

The issue occurs because of the assumptions made [here](https://github.com/emberjs/data/blame/47a71ca1538ba9e2d7dfa01bf048a2db897bdf5f/packages/record-data/addon/-private/record-data.ts#L285..L290), when `didCommit` is invoked after save1 completes.  The code (incorrectly) assumes that:
 * All in-flight attributes have been seen by the server, 
 * Therefore they are accounted for in the `data` returned by the server, and 
 * Therefore they are merged into `remoteattrs` and dropped from `inflightattrs`, causing them to no longer visible to the app.

To be honest, it's probably quite difficult to solve this issue in the general case, because the concurrent save requests can race one another and succeed (or fail!) individually in any order.  Furthermore, the timing of requests resolving or rejecting locally does not necessarily imply that those requests were received/processed in the same order on the server, since concurrent requests can theoretically race in all phases of the request lifecycle, including even in the network transporting the response back to the browser.

Since the typical behavior is that earlier requests are processed and complete before latest requests, I suspect the preferred behavior here would be for Ember Data to somehow track the latest `.save()` request and only clear `inflightattrs` once the final request issued resolves.  It would also need to figure out how to update the record state appropriately as older requests resolve/reject before the final one, and it would still need to make some assumptions on the record state if an older request ever resolves/rejects after the final one.

Anyway, unless and until a solution like the above is implemented, the most effective workaround for this solution is to employ some form of "save throttling" as described [above](#issuecomment-190261816).  This avoids not only the incorrect behavior of Ember Data, but also the non-deterministic effects that concurrent requests can produce in general.

**Comment by robbytx** [comment]
Here is a modern version of the workaround for this issue:
```js
/**
 * This function can be used to override the `save` method in subclasses of Model from '@ember-data/model', like:
 * <code>
 * class MyModel extends Model {
 *   ...
 *   save = serialSave;
 * }
 * </code>
 *
 * This function may be invoked for a particular record as many times as desired, and it will ensure that only a
 * single execution of .save() is pending at any given time, to workaround https://github.com/emberjs/data/issues/4195
 * as well as non-deterministic outcomes that might result from concurrent save requests.
 * 
 * Although the behavior of this function can result in coalescing of multiple .save() requests, it specifically does
 * not detect and elide a deferred .save() request that would appear to have no effect.  Users wishing to implement
 * this optimization should check .isSaving and other Model APIs to check if there are pending changes before calling this method.   
 *
 * @returns {Promise} containing the results of the next .save() operation that includes the current attributes.
 */
function serialSave() {
  // This function performs a single .save() execution, whose promise is tracked as this.pendingSave
  const performSave = () => {
    const superSave = Object.getPrototypeOf(Object.getPrototypeOf(this)).save; // workaround for super.save() in anonymous function
    return this.pendingSave = superSave.call(this).finally(() => this.pendingSave = null);
  };
  if (this.deferredSave) {
    // Another .save() is already scheduled to execute after this.pendingSave resolves/rejects
    return this.deferredSave;
  } else if (this.pendingSave) {
    // Create a single Promise that will perform a subsequent .save() once pendingSave resolves/rejects
    return this.deferredSave = new Promise((resolve, reject) => {
      this.pendingSave.finally(() => {
        this.deferredSave = null;
        return performSave().then(resolve, reject);
      })
    });
  } else {
    // No .save() is currently executing, so proceed immediately!
    return performSave();
  }
}
```

**Comment by robbytx** [comment]
Related to this, I've realized that modern versions of Ember Data even discard in-flight attributes when a second save() request is issued ([link](https://github.com/emberjs/data/blob/v4.8.4/packages/record-data/addon/-private/record-data.ts#L257)):
```js
// 0. Create record with specific state
const record = await store.createRecord('type', { counter: 0, foo: null });

// 1. First update
record.counter = 1;
const save1 = record.save(); // sends { attributes: { counter: 1, foo: null } } as expected

// 2. Second update (issued in separate runloop, but before save1 resolves):
record.foo = 'bar';
const save2 = record.save(); // sends { attributes: { counter: 0, foo: 'bar' } } <-- counter has reverted to original state!
record.counter === 0; // counter is also reverted on the model
```

This was surprising to me, but the old InternalModel code path had similar logic ([link](https://github.com/emberjs/data/blob/v3.4.4/addon/-private/system/model/internal-model.js#L785).  Interestingly, however, in Ember <3.28, it seems that there were some key differences:

1.  The snapshot to save was created before the changed attributes were flushed. ([link](https://github.com/emberjs/data/blob/v3.4.4/addon/-private/system/store.js#L1992))
2. The snapshot used EmberObject `get`, which additionally cached the value of the attributes ([link](https://github.com/emberjs/data/blob/v3.4.4/addon/-private/system/snapshot.js#L94)) until it was notified of an update.

It's unclear to me precisely when the behavior changed, but I suspect it was during the Ember 3.28 release when the record data and other code paths were activated by default.

My only goal in noting this is to clarify further the behavior of Ember Data in response to concurrent saves.  I'm exploring some options in my own app to see if I can develop a coherent strategy for handling this.

**Comment by runspired** [comment]
It looks like the regression @robbytx noted occurred in 4.12.

We're adding a fix to 4.12+ that makes concurrent saves *mostly* behave as expected, though there is a subtle bug potential in that inflight state is updated by each subsequent request and thus the first request to resolve assumes all inflight state is now correctly saved.

The better approach for apps that do have lots of concurrent saves and need strong guarantees is to use the request manager to keep track of the precise changes associated to each request, and to use `PATCH` instead of `PUT` API behaviors.

https://github.com/emberjs/data/pull/9012
